File: sdk/typescript/src/encryption.ts
/**
 * Prismer SDK — E2E Encryption
 *
 * Industry-standard end-to-end encryption for IM messages.
 *
 * - Per-conversation symmetric key: AES-256-GCM
 * - Key exchange: ECDH P-256
 * - Master key derivation: PBKDF2-SHA256
 * - Key storage: Encrypted with master key
 * - Runtime: Web Crypto API (browser) / node:crypto (Node.js)
 *
 * The server only sees ciphertext — it cannot decrypt message content.
 */

// ============================================================================
// Crypto abstraction — works in both browser and Node.js
// ============================================================================

interface CryptoBackend {
  generateKeyPair(): Promise<{ publicKey: JsonWebKey; privateKey: JsonWebKey }>;
  deriveSharedSecret(privateKey: JsonWebKey, publicKey: JsonWebKey): Promise<ArrayBuffer>;
  deriveMasterKey(passphrase: string, salt: Uint8Array): Promise<ArrayBuffer>;
  encrypt(key: ArrayBuffer, plaintext: string): Promise<string>;
  decrypt(key: ArrayBuffer, ciphertext: string): Promise<string>;
  randomBytes(length: number): Uint8Array;
}

function getSubtleCrypto(): SubtleCrypto {
  if (typeof globalThis.crypto?.subtle !== 'undefined') {
    return globalThis.crypto.subtle;
  }
  // Node.js
  try {
    const { webcrypto } = require('node:crypto');
    return webcrypto.subtle;
  } catch {
    throw new Error('No SubtleCrypto available. Requires browser or Node.js 16+.');
  }
}

function getRandomValues(arr: Uint8Array): Uint8Array {
  if (typeof globalThis.crypto?.getRandomValues !== 'undefined') {
    return globalThis.crypto.getRandomValues(arr);
  }
  try {
    const { webcrypto } = require('node:crypto');
    return webcrypto.getRandomValues(arr);
  } catch {
    throw new Error('No crypto.getRandomValues available.');
  }
}

const subtle = (): SubtleCrypto => getSubtleCrypto();

const PBKDF2_ITERATIONS = 100_000;
const SALT_LENGTH = 16;
const IV_LENGTH = 12; // AES-GCM recommended
const KEY_LENGTH = 256; // AES-256

// ============================================================================
// E2E Encryption
// ============================================================================

/**
 * E2EEncryption manages per-conversation symmetric keys and
 * encrypts/decrypts message content using AES-256-GCM.
 *
 * Usage:
 *   const e2e = new E2EEncryption();
 *   await e2e.init('user-passphrase');
 *   const ciphertext = await e2e.encrypt('conv-123', 'Hello!');
 *   const plaintext = await e2e.decrypt('conv-123', ciphertext);
 */
export class E2EEncryption {
  private masterKey: CryptoKey | null = null;
  private keyPair: CryptoKeyPair | null = null;
  private sessionKeys = new Map<string, CryptoKey>(); // conversationId → AES key
  private salt: Uint8Array | null = null;

  /**
   * Initialize encryption with user passphrase.
   * Derives a master key via PBKDF2 and generates an ECDH key pair.
   */
  async init(passphrase: string): Promise<void> {
    // Generate or load salt
    this.salt = getRandomValues(new Uint8Array(SALT_LENGTH));

    // Derive master key from passphrase
    const passphraseKey = await subtle().importKey(
      'raw',
      new TextEncoder().encode(passphrase),
      'PBKDF2',
      false,
      ['deriveKey'],
    );

    this.masterKey = await subtle().deriveKey(
      {
        name: 'PBKDF2',
        salt: this.salt as BufferSource,
        iterations: PBKDF2_ITERATIONS,
        hash: 'SHA-256',
      },
      passphraseKey,
      { name: 'AES-GCM', length: KEY_LENGTH },
      false,
      ['encrypt', 'decrypt'],
    );

    // Generate ECDH key pair for key exchange
    this.keyPair = await subtle().generateKey(
      { name: 'ECDH', namedCurve: 'P-256' },
      true,
      ['deriveKey'],
    );
  }

  /**
   * Export public key for sharing with conversation peers.
   */
  async exportPublicKey(): Promise<JsonWebKey> {
    if (!this.keyPair) throw new Error('E2E not initialized. Call init() first.');
    return subtle().exportKey('jwk', this.keyPair.publicKey);
  }

  /**
   * Derive a shared session key for a conversation using ECDH.
   * Call this with each peer's public key.
   */
  async deriveSessionKey(conversationId: string, peerPublicKey: JsonWebKey): Promise<void> {
    if (!this.keyPair) throw new Error('E2E not initialized. Call init() first.');

    const importedPeerKey = await subtle().importKey(
      'jwk',
      peerPublicKey,
      { name: 'ECDH', namedCurve: 'P-256' },
      false,
      [],
    );

    const sessionKey = await subtle().deriveKey(
      { name: 'ECDH', public: importedPeerKey },
      this.keyPair.privateKey,
      { name: 'AES-GCM', length: KEY_LENGTH },
      false,
      ['encrypt', 'decrypt'],
    );

    this.sessionKeys.set(conversationId, sessionKey);
  }

  /**
   * Set a pre-shared session key for a conversation.
   * Useful when the key is exchanged out-of-band or derived from a group key.
   */
  async setSessionKey(conversationId: string, rawKey: ArrayBuffer): Promise<void> {
    const key = await subtle().importKey(
      'raw',
      rawKey,
      { name: 'AES-GCM', length: KEY_LENGTH },
      false,
      ['encrypt', 'decrypt'],
    );
    this.sessionKeys.set(conversationId, key);
  }

  /**
   * Generate a random session key for a conversation.
   * Returns the raw key bytes for sharing with peers.
   */
  async generateSessionKey(conversationId: string): Promise<ArrayBuffer> {
    const key = await subtle().generateKey(
      { name: 'AES-GCM', length: KEY_LENGTH },
      true,
      ['encrypt', 'decrypt'],
    );
    this.sessionKeys.set(conversationId, key);
    return subtle().exportKey('raw', key);
  }

  /**
   * Encrypt plaintext for a conversation.
   * Returns base64-encoded ciphertext with prepended IV.
   */
  async encrypt(conversationId: string, plaintext: string): Promise<string> {
    const key = this.sessionKeys.get(conversationId);
    if (!key) throw new Error(`No session key for conversation ${conversationId}. Call deriveSessionKey() first.`);

    const iv = getRandomValues(new Uint8Array(IV_LENGTH));
    const encoded = new TextEncoder().encode(plaintext);

    const ciphertext = await subtle().encrypt(
      { name: 'AES-GCM', iv: iv as BufferSource },
      key,
      encoded,
    );

    // Prepend IV to ciphertext and base64-encode
    const combined = new Uint8Array(iv.length + ciphertext.byteLength);
    combined.set(iv, 0);
    combined.set(new Uint8Array(ciphertext), iv.length);

    return arrayBufferToBase64(combined.buffer);
  }

  /**
   * Decrypt ciphertext from a conversation.
   * Expects base64-encoded data with prepended IV.
   */
  async decrypt(conversationId: string, ciphertext: string): Promise<string> {
    const key = this.sessionKeys.get(conversationId);
    if (!key) throw new Error(`No session key for conversation ${conversationId}. Call deriveSessionKey() first.`);

    const combined = base64ToArrayBuffer(ciphertext);
    const iv = combined.slice(0, IV_LENGTH);
    const data = combined.slice(IV_LENGTH);

    const decrypted = await subtle().decrypt(
      { name: 'AES-GCM', iv: new Uint8Array(iv) },
      key,
      data,
    );

    return new TextDecoder().decode(decrypted);
  }

  /**
   * Check if a session key exists for a conversation.
   */
  hasSessionKey(conversationId: string): boolean {
    return this.sessionKeys.has(conversationId);
  }

  /**
   * Remove session key for a conversation.
   */
  removeSessionKey(conversationId: string): void {
    this.sessionKeys.delete(conversationId);
  }

  /**
   * Clear all keys and reset state.
   */
  destroy(): void {
    this.masterKey = null;
    this.keyPair = null;
    this.sessionKeys.clear();
    this.salt = null;
  }
}

// ============================================================================
// Helpers
// ============================================================================

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  if (typeof btoa !== 'undefined') {
    // Browser
    const bytes = new Uint8Array(buffer);
    let binary = '';
    for (let i = 0; i < bytes.byteLength; i++) {
      binary += String.fromCharCode(bytes[i]);
    }
    return btoa(binary);
  }
  // Node.js
  return Buffer.from(buffer).toString('base64');
}

function base64ToArrayBuffer(base64: string): ArrayBuffer {
  if (typeof atob !== 'undefined') {
    // Browser
    const binary = atob(base64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes.buffer;
  }
  // Node.js
  const buf = Buffer.from(base64, 'base64');
  return buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength);
}

File: sdk/prismer-cloud/python/prismer/encryption.py
"""E2E Encryption — AES-256-GCM with ECDH P-256 key exchange.

Interoperable with TypeScript, Go, and Rust SDKs:
- Master key: PBKDF2-SHA256 (100k iterations), random 16-byte salt (per-user)
- Session keys: AES-256-GCM per conversation
- Ciphertext format: base64(12-byte-IV + ciphertext)
"""

import base64
import hashlib
import os
from typing import Dict, Optional, Tuple

from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes, serialization


PBKDF2_ITERATIONS = 100_000
KEY_LENGTH = 32  # AES-256
IV_LENGTH = 12   # GCM nonce
SALT_LENGTH = 16


class E2EEncryption:
    """E2E encryption manager — AES-256-GCM + ECDH P-256."""

    def __init__(self) -> None:
        self._master_key: Optional[bytes] = None
        self._salt: Optional[bytes] = None
        self._session_keys: Dict[str, bytes] = {}
        self._private_key: Optional[ec.EllipticCurvePrivateKey] = None
        self._public_key_bytes: Optional[bytes] = None

    def init(self, passphrase: str, salt: Optional[str] = None) -> None:
        """Initialize with passphrase — derives master key via PBKDF2.

        Args:
            passphrase: User passphrase for master key derivation.
            salt: Optional Base64-encoded salt. If omitted, a random 16-byte salt
                  is generated. Store via export_salt() to re-derive the same key.
        """
        self._salt = base64.b64decode(salt) if salt else os.urandom(SALT_LENGTH)
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=KEY_LENGTH,
            salt=self._salt,
            iterations=PBKDF2_ITERATIONS,
        )
        self._master_key = kdf.derive(passphrase.encode("utf-8"))

        # Generate ECDH P-256 keypair
        self._private_key = ec.generate_private_key(ec.SECP256R1())
        self._public_key_bytes = self._private_key.public_key().public_bytes(
            serialization.Encoding.X962,
            serialization.PublicFormat.UncompressedPoint,
        )

    @property
    def is_initialized(self) -> bool:
        return self._master_key is not None

    def export_salt(self) -> str:
        """Export the salt as Base64 for persistent storage."""
        if self._salt is None:
            raise RuntimeError("Call init() first")
        return base64.b64encode(self._salt).decode("ascii")

    def export_public_key(self) -> Optional[str]:
        """Export public key as base64 for sharing with peers."""
        if self._public_key_bytes is None:
            return None
        return base64.b64encode(self._public_key_bytes).decode("ascii")

    def derive_session_key(self, conversation_id: str, peer_public_key_b64: str) -> bytes:
        """Derive a shared session key via ECDH from a peer's public key."""
        if self._private_key is None:
            raise RuntimeError("Call init() first")
        peer_bytes = base64.b64decode(peer_public_key_b64)
        peer_key = ec.EllipticCurvePublicKey.from_encoded_point(ec.SECP256R1(), peer_bytes)
        shared_key = self._private_key.exchange(ec.ECDH(), peer_key)
        # Use first 32 bytes of SHA-256(shared_secret) as session key
        session_key = hashlib.sha256(shared_key).digest()
        self._session_keys[conversation_id] = session_key
        return session_key

    def set_session_key(self, conversation_id: str, key: bytes) -> None:
        """Set a pre-shared session key for a conversation."""
        if len(key) != KEY_LENGTH:
            raise ValueError(f"Session key must be {KEY_LENGTH} bytes")
        self._session_keys[conversation_id] = key

    def generate_session_key(self, conversation_id: str) -> bytes:
        """Generate a random session key for a conversation."""
        key = os.urandom(KEY_LENGTH)
        self._session_keys[conversation_id] = key
        return key

    def has_session_key(self, conversation_id: str) -> bool:
        return conversation_id in self._session_keys

    def encrypt(self, conversation_id: str, plaintext: str) -> str:
        """Encrypt plaintext. Returns base64(IV + ciphertext)."""
        key = self._session_keys.get(conversation_id)
        if key is None:
            raise KeyError(f"No session key for conversation: {conversation_id}")
        iv = os.urandom(IV_LENGTH)
        aesgcm = AESGCM(key)
        ciphertext = aesgcm.encrypt(iv, plaintext.encode("utf-8"), None)
        return base64.b64encode(iv + ciphertext).decode("ascii")

    def decrypt(self, conversation_id: str, encrypted: str) -> str:
        """Decrypt base64(IV + ciphertext)."""
        key = self._session_keys.get(conversation_id)
        if key is None:
            raise KeyError(f"No session key for conversation: {conversation_id}")
        combined = base64.b64decode(encrypted)
        if len(combined) < IV_LENGTH + 1:
            raise ValueError("Ciphertext too short")
        iv = combined[:IV_LENGTH]
        ciphertext = combined[IV_LENGTH:]
        aesgcm = AESGCM(key)
        plaintext = aesgcm.decrypt(iv, ciphertext, None)
        return plaintext.decode("utf-8")
